rsx/rsx_lib_gl.cpp

```diff
@@ -1834,12 +1796,22 @@ static void vertex_preprocessing(
       CommandVertex *v,
       unsigned count,
       GLenum mode,
-      SemiTransparencyMode stm)
+      SemiTransparencyMode stm,
+	  bool mask_test,
+	  bool set_mask)
 {
    if (!renderer)
       return;
 
    bool is_semi_transparent = v[0].semi_transparent == 1;
+   bool is_textured         = v[0].texture_blend_mode != 0;
+   /* Textured semi-transparent polys can contain opaque texels (when
+    * bit 15 of the color is set to 0). Therefore they're drawn twice,
+    * once for the opaque texels and once for the semi-transparent
+    * ones. Only untextured semi-transparent triangles don't need to be
+    * drawn as opaque. */
+   bool is_opaque = !is_semi_transparent || is_textured;
```

Comment:
Primitives which have mask test aren't necessarily opaque. They depend on the latest mask bit in the framebuffer, so it needs to be considered as blending.

Reply:
Right, semi-transparent primitives with mask bits are also handled.